nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,304 @@
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_CONE_CLIENT_2_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_SHARED_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=False,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_SHARED_CLIENT_1_histogram_check,
+            ),
+        ])
+    ),
+]
+
+
+async def get_batch_histogram(
+    target: TargetMachine, mesh_node: Node
+) -> HistogramResult:
+    print(f"Running batch observer for target: {target.tag}")
+
+    await asyncio.sleep(2)
+    async with BatchObserver(target).run() as bo:
+        await asyncio.sleep(20)
+
+        local_path = f"./{target.tag}.pcap"
+        remote_path = local_path
+
+        subprocess.run([
+            "docker",
+            "cp",
+            target.container_id + ":" + remote_path,
+            local_path,
+        ])
+
+        # talking in milliseconds
+        hs_in = bo.get_histogram(
+            10,
+            500,
+            ObservationTarget(Direction.Incoming, mesh_node.ip_addresses[0]),
+        )
+        hs_out = bo.get_histogram(
+            10,
+            500,
+            ObservationTarget(Direction.Outgoing, mesh_node.ip_addresses[0]),
+        )
+        os.unlink(local_path)
+
+        return HistogramResult(hs_in, hs_out)
+
+
+@pytest.mark.asyncio
+async def test_histogram():
+    bo = Histogram()
+    for _ in range(10):
+        bo.add_value(2)
+        bo.add_value(3)
+
+    for _ in range(50):
+        bo.add_value(4)
+
+    bo.add_value(9)
+
+    assert bo.get(10, 1) == [0, 0, 10, 10, 50, 0, 0, 0, 0, 1]
+
+
+# nodes that will capture the traffic
+capture_nodes: List[ConnectionTag] = [
+    ConnectionTag.DOCKER_CONE_CLIENT_1,
+    ConnectionTag.DOCKER_CONE_CLIENT_2,
+    ConnectionTag.DOCKER_SHARED_CLIENT_1,
+]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("client_params", [CLIENTS_NO_BATCHING, CLIENTS_BATCHING])
+async def test_direct_batching(client_params: Any) -> None:
+    async with AsyncExitStack() as exit_stack:
+        for setup_params in client_params:
+            # This testcase initiates 2 meshnet peers, and after a small delay initiates the third peer.
+            # The reason for this is to misalign their keepalive echo requests and observe batching in action
+            # The third peer has naturally aligned keepalives even if with no batching as it received meshmap
+            # with both peers at once and added the keys
+
+            setup_params = setup_params[0][0]  # TODO: Pytest, wtf
```

Comment:
So I was a bit frustrated because Pytest has no worries iteracting over a singular SetupParameters by symply accessing `setup_parameters` however once I convert that to an array suddenly I need double the indirection which does'nt really make any sense to me, but of course, I will explain it further, good point :)